pydantic/plugin/_loader.py

```diff
@@ -18,36 +18,41 @@
 
 PYDANTIC_ENTRY_POINT_GROUP: Final[str] = 'pydantic'
 
-_plugins: dict[str, PydanticPluginProtocol] = {}
-_all_plugins_loaded: bool = False
+# cache of plugins
+_plugins: dict[str, PydanticPluginProtocol] | None = None
+# return no plugins while loading plugins to avoid recursion and errors while import plugins
+# this means that if plugins use pydantic
+_loading_plugins: bool = False
 
 
 def get_plugins() -> Iterable[PydanticPluginProtocol]:
     """Load plugins for Pydantic.
 
     Inspired by: https://github.com/pytest-dev/pluggy/blob/1.3.0/src/pluggy/_manager.py#L376-L402
     """
-    global _plugins, _all_plugins_loaded
-    if _all_plugins_loaded:
-        return _plugins.values()
-
-    _all_plugins_loaded = True
-
-    for dist in importlib_metadata.distributions():
-        for entry_point in dist.entry_points:
-            if entry_point.group != PYDANTIC_ENTRY_POINT_GROUP:
-                continue
-            if entry_point.value in _plugins:
-                continue
-            try:
-                _plugins[entry_point.value] = entry_point.load()
-            except (ImportError, AttributeError) as e:
-                _all_plugins_loaded = False
-                error_type = e.__class__.__name__
-                warnings.warn(
-                    f'{error_type} while loading the `{entry_point.name}` Pydantic plugin, this could be caused '
-                    f'by a circular import issue (e.g. the plugin importing Pydantic), Pydantic will attempt to '
-                    f'import this plugin again each time a Pydantic validators is created. {e}'
-                )
+    global _plugins, _loading_plugins
+    if _loading_plugins:
+        # this happens when plugins themselves use pydantic, we return no plugins
+        return {}.values()
```

Comment:
Just an empty tuple should be enough as `Iterable`

Reply:
yes, but I wanted to keep the type exactly the same.

Reply:
I'll add a comment.

Reply:
I think you're right, it doesn't need to be dict values.